src/modules/logger.js

```diff
@@ -0,0 +1,205 @@
+    /* Verify any path to directory is writable */
+
+    let newLogFilePathGTP = setting.get('gtp.console_log_path')
+    let loggingEnabled = setting.get('gtp.console_log_enabled')
+    // For GUI don't show a warning when logging is disabled AND the path is empty
+    if (newLogFilePathGTP === null) {
+        if (!loggingEnabled) {
+            staleLogFilePath = false
+            return true
+        } else {
+            staleLogFilePath = false
+            return false
+        }
+    }
+
+    let fileStats = null
+    try {
+        fileStats = fs.statSync(newLogFilePathGTP)
+    } catch (err) {}
+
+    // if fileStats null, path doesnt exist
+    if (fileStats != null) {
+        if (fileStats.isDirectory()) {
+            try {
+                fs.accessSync(newLogFilePathGTP, fs.W_OK)
+                staleLogFilePath = true
+                return true
+            } catch (err) {}
+        }
+        // Path exists, either no write permissions to directory, or path is not a directory
+        staleLogFilePath = true
+        return false
+    } else {
+        staleLogFilePath = false
+        return false
+    }
+}
+
+exports.validateLoggerSettings = function() {
+    if (!exports.validLogFilePathGTP()) {
+        let loggingEnabled = setting.get('gtp.console_log_enabled')
+        if (loggingEnabled) {
+            dialog.showMessageBox([
+                'You have an invalid log folder for GTP console logging in your settings.\n\n',
+                'Please make sure the log directory is valid & writable or disable GTP console logging.',
+                ].join(''), 'warning'
+            )
+        }
+        return false
+    } else {
+        return true
+    }
+}
+
+exports.loadOnce = function() {
+    if (logger == null) {
+        if(setting.get('gtp.console_log_timestamps')) {
+            logger = winston.createLogger({
+                format: winston.format.combine(
+                    winston.format.timestamp({format: 'YYYY-MM-DD HH:mm:ss.SSS'}),
+                    winston.format.printf(info => {
+                        return `\[${info.timestamp}\] ${info.message}`;
+                    })),
+                handleExceptions: false,
+                exitOnError: false});
+        } else {
+            logger = winston.createLogger({
+                format: winston.format.combine(
+                    winston.format.printf(info => {
+                        return `${info.message}`;
+                    })),
+                handleExceptions: false,
+                exitOnError: false});
+        }
+    }
+    if (exports.validateLoggerSettings()) {
+        exports.updateLogFilePathGTP()
+    }
+}
+
+exports.rotateLog = function() {
+    // On next engine attach, we will use a new log file
+    logFileNameGTP = null;
+}
+
+exports.unload = function() {
+    try {
+        logger.close()
+    } catch (err) {}
+}
+
+exports.close = function() {
+    try {
+        logger.close()
```

Comment:
What does `logger.close()` do exactly? I can't find this method in [winston's documentation](https://github.com/winstonjs/winston). Can you point me to some reference?

Reply:
`close()` calls `clear()`, which closes all the logger transport streams -- all in all the purpose of it in our code is to **make sure the data is flushed before we quit**.

There isn't  thorough documentation for all of winston, so I had to go through winston's code to answer you more definitively:

`close()` calls `clear()`, which closes all the streams with Node's `stream.unpipe()` -- from my understanding, removing this writable stream with unpipe makes sure any data in the stream's buffer is flushed to disk before we actually quit the application.

**The only reason for using `close()` rather than `clear()` is it's easier to understand than `clear()` from first glance** (I've seen it used interchangeably with `clear()` in winston examples).  `clear()` _sounds_ like it is clearing the buffer, but that is not what it does (terrible naming IMO) -- it just removes all the transport streams for the logger instance using `unpipe()`.

`.close()` real intended use seems to be for the case when you have a container of different winston logger instances, then you can use events to manage them: `close()` fires a 'close' event an event to let the manager know we did a `clear()` on that logger instance

**On a related note**: at the moment we don't actually bother to remove the logger until we need to actually start a new logging file with `updateLoggerPath()`. There is a very good reason for this: the logger is async, so its always possible to be writing data past a Sabaki call for a new board with `loadGameTrees()`, especially if the user didn't manually detach the engines first, so we need to keep the current logger open as long as possible in that eventuality. (In testing I saw this happen, which is why it is like it is rather than using a much more complicated callback system).

Here are the code line references in the source to `close` & `clear()`:

`close()` calls `clear()`
https://github.com/winstonjs/winston/blob/64344d12eace394d9e7c8bf926f9f2663efc8640/lib/winston/logger.js#L348

`clear()` calls `unpipe()`
https://github.com/winstonjs/winston/blob/64344d12eace394d9e7c8bf926f9f2663efc8640/lib/winston/logger.js#L338